tests/background/tab-state-test.js

```diff
@@ -132,23 +132,126 @@ describe('TabState', () => {
+
+    it('cancels the first query (during waiting period) when the service is called two consecutive times for the same tab', async () => {
+      const testValue = 42;
+      getStub.resolves(testValue);
+      const tabState = new TabState({ 1: { state: states.ACTIVE } });
+
+      const promise1 = tabState.updateAnnotationCount(1, 'foobar.com');
+      clock.tick(800); // promise 1 is still waiting when promise2 is called
+      const promise2 = tabState.updateAnnotationCount(1, 'foobar.com');
+      clock.tick(2300);
+      await promise1;
+      assert.strictEqual(tabState.getState(1).annotationCount, 0);
+
+      await promise2;
+      assert.calledOnce(getStub);
+      assert.strictEqual(tabState.getState(1).annotationCount, testValue);
+    });
+
+    it('waits for a maximum of 3000 ms when a number of concurrent requests', async () => {
+      const testValue = 42;
+      getStub.resolves(testValue);
       const tabState = new TabState({ 1: { state: states.ACTIVE } });
-      return tabState.updateAnnotationCount(1, 'foobar.com').then(() => {
-        assert.called(getStub);
-        assert.equal(tabState.getState(1).annotationCount, testValue);
+
+      const promise1 = tabState.updateAnnotationCount(1, 'foobar.com');
+      const promise2 = tabState.updateAnnotationCount(1, 'foobar.com');
+      const promise3 = tabState.updateAnnotationCount(1, 'foobar.com');
+      const promise4 = tabState.updateAnnotationCount(1, 'foobar.com');
+      const promise5 = tabState.updateAnnotationCount(1, 'foobar.com');
+      clock.tick(3010); // promise 1-4 are cancelled, while promise 4 is resolved in no more than 3000 ms
+      await promise1;
+      await promise2;
+      assert.strictEqual(tabState.getState(1).annotationCount, 0);
+      await promise3;
+      await promise4;
+      await promise5;
+      assert.calledOnce(getStub);
+      assert.strictEqual(tabState.getState(1).annotationCount, testValue);
+    });
+
+    it('cancels the first query (during the fetch request) when the service is called two consecutive times for the same tab', async () => {
+      const testValue = 42;
+
+      // Takes 2000ms in returning a the response
+      getStub.returns(
+        new Promise(resolve => setTimeout(() => resolve(testValue), 2000))
+      );
+
+      const tabState = new TabState({ 1: { state: states.ACTIVE } });
+
+      const promise1 = tabState.updateAnnotationCount(1, 'foobar.com');
+      clock.tick(1010); // promise1 finished waiting and it is fetching the request
+      const promise2 = tabState.updateAnnotationCount(1, 'foobar.com');
+      clock.tick(1010 + 2000);
+
+      await promise1;
+      assert.calledTwice(getStub); // request is not cancelled
+      assert.equal(tabState.getState(1).annotationCount, 0);
+      await promise2;
+      assert.equal(tabState.getState(1).annotationCount, testValue);
+    });
+
+    it('processes two consecutive request to the service if the requests are for different tabs', async () => {
```

Comment:
I know what you mean by this description, but I think perhaps the wording of this test description could do with an adjustment. I'm still trying to think of the best way to express this.